core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -90,39 +90,50 @@ public void onExpiration() {
      */
     @Override
     public void onComplete() {
+        // We are utilizing lock so that onComplete doesn't do a dirty read for global variables -
+        // partitionsAcquired and partitionsAlreadyFetched, since these variables can get updated in a different tryComplete thread.
+        lock.lock();
         log.trace("Completing the delayed share fetch request for group {}, member {}, "
             + "topic partitions {}", shareFetchData.groupId(), shareFetchData.memberId(),
             partitionsAcquired.keySet());
 
-        if (shareFetchData.future().isDone())
-            return;
+        try {
+            if (shareFetchData.future().isDone())
+                return;
```

Comment:
yeah, I agree that's a super corner case scenario, but definitely possible. I have pushed a fix for it. Thanks for pointing it out.

Reply:
Hmm, could shareFetchData.future().isDone() be true inside onComplete()? We complete the future only after DelayedOperation.completed is set to true. After that point, onComplete() is not expected to be called again.

Reply:
So, if we have 2 different keys corresponding to a ShareFetch request, it could be a case that for one of those keys, we get a `checkAndComplete` call which could result in completing the share fetch request. Now when the purgatory entry corresponding to the other key could timeout/have `checkAndComplete` triggered, when the code reaches onComplete, the share fetch request's future was already complete, so it would hit `shareFetchData.future().isDone()` and return true.

Reply:
`onComplete` is always called through `forceComplete`, right? So, only one thread could ever call `onComplete`.

```
    public boolean forceComplete() {
        if (completed.compareAndSet(false, true)) {
            // cancel the timeout timer
            cancel();
            onComplete();
            return true;
        } else {
            return false;
        }
    }

```

Reply:
Hi @junrao , you're right. There was a gap in my understanding of purgatory operation where I thought the the copy of the operation goes to multiple watch keys used for that operation, but this line in documentation cleared it out.
```
Note that a delayed operation can be watched on multiple keys. It is possible that an operation is completed after it has been added to the watch list for some, but not all the keys. In this case, the operation is considered completed and won't be added to the watch list of the remaining keys. The expiration reaper thread will remove this operation from any watcher list in which the operation exists.
```
Hence, I've removed the mentioned condition from the code now. Thanks!

Reply:
> In this case, the operation is considered completed and won't be added to the watch list of the remaining keys. 

This comment seems inaccurate. Currently, we add a delayed operation to either all keys or none of the keys. Nevertheless, the property that onComplete() is only called once when the operation is not completed still holds. It would be useful to fix the comment.

Reply:
sure, I'll cover it as part of JIRA https://issues.apache.org/jira/browse/KAFKA-18027 in the near future..